packages/web-frontend/src/routes/CreateCardComponent.svelte

```diff
@@ -0,0 +1,112 @@
+
+    type Loading = {
+        state: 'loading';
+    }
+
+    type Error = {
+        state: 'error';
+        message?: string;
+    }
+
+    type ClozeData = {
+        type: 'cloze';
+        front: string;
+    }
+
+    type QaData = {
+        type: 'qa';
+        front: string;
+        back: string;
+    }
+
+    type CreateCardData = ClozeData | QaData;
+
+    type Data = {
+        state: 'data';
+        data: CreateCardData;
+    }
+
+    type UiState = Loading | Error | Data;
+
+    // Note, we'd ideally have this is some sort of generic type and the component would define
+    // something like UiState<SpecificComponentDataType>
+    export let state: UiState = { state: 'loading' };
+
+    // events are things the user does with the component, for example ClickSubmit or ChangeFront or ChangeBack
+    type ClickSubmit = {
+        front: string;
+        back: string;
+    }
+
+    type ChangeFront = {
+        newValue: string;
+    }
+
+    type ChangeBack = {
+        newValue: string;
+    }
+
+    type Event = ClickSubmit | ChangeFront | ChangeBack;
+
+    // events are exposed up the tree to an appropriate handler via an eventSink prop
+    export let eventSink: (event: Event) => void = defaultEventSink;
+
+    function defaultEventSink(event: Event) {
+        console.log('Unhandled event', event);
+    }
+  
+</script>
+
+<!-- Now we produce HTML based on the uiState field, and emit events based on DOM interactions -->
+
+<!-- We'd *love* to have Algebraic Data Types here with exhaustive type matching in a `when`/`switch`/`match` construct
+    But alas, no such thing exists. So we manually check via if/else, which is unfortunately much more error-prone.
+-->
+{#if state.state === 'loading'}
+    <div>Loading...</div>
+{:else if state.state === 'error'}
+    <div>Error: {state.message}</div>
+{:else if state.state === 'data'}
+    <div></div>
+    <!-- Again, we'd love to have ADTs and exhaustive matching checks here, but c'est la vie -->
+    {#if state.data.type === 'cloze'}
+    <div>Cloze card</div>
+        <!-- <CreateClozeCardComponent {state.data} {eventSink} /> -->
+    {:else if state.data.type === 'qa'}
+    <div>QA Card</div>
+        <!-- <CreateQaCardComponent {state.data} {eventSink} /> -->
+    {/if}
+
+        <div>
+            <input type="text" value={state.data.front} on:input={(e) => eventSink({ newValue: e.target.value })} />
+            <button on:click={() => eventSink({ newValue: state.data.front })}>Submit</button>
+        </div>
+    <!-- {:else if state.data.front && state.data.back}
+        <div>
+            <input type="text" value={state.data.front} on:input={(e) => eventSink({ newValue: e.target.value })} />
+            <input type="text" value={state.data.back} on:input={(e) => eventSink({ newValue: e.target.value })} />
+            <button on:click={() => eventSink({ front: state.data.front, back: state.data.back })}>Submit</button>
+        </div>
+    {/if} -->
+    <div>
```

Comment:
I'm confused by this <div>, why does it exist when there's the div up above which it almost the same, except this div has an extra <input> field?

I see you have the `<!-- <CreateClozeCardComponent {state.data} {eventSink} /> -->` and `<!-- <CreateQaCardComponent {state.data} {eventSink} /> -->` up above... are these 2 existing divs supposed to be handling the 2 different card types? If so, then shouldn't they be in the `{#if state.data.type === 'cloze'}` and `{:else if state.data.type === 'qa'}` blocks up above?

Reply:
1) What is this?

This is a remnant of one of the approaches I considered for the UI. The question is whether to split this component into two very simple subcomponents (one for creating a cloze card, one for front back) or merge those things into a single component. The two subcomponents are *pretty* similar, only differing by a text field and some notion of "type", so you could pretty easily just leave it in a single component and not add extra types/api overhead. I'm still undecided on which one is preferable, and left this here while I worked through the other options a little.

2) Why isn't it `{#if state.data.type === 'cloze'}`?

We could do this, but I don't love it, conceptually. It's operating at a different level of abstraction compared with the top-level ui state check (top level states being `loading`, `data`, `error`). Not fundamentally opposed, but I find it easier to read/parse when things at the same indentation level are at the same abstraction level. Additionally, a check of `loading`, `data`, `error` would make for much more structural consistency across components.

Ultimately, the difference is like a 20% thing not an 80% thing. I think this approach is 20% better than checking inside the `data` object at the top level of a component, but it'd be fine to go another direction if you have a strong preference.